crates/integration_test/src/tests/readonly_paths/readonly_paths.rs

```diff
@@ -0,0 +1,275 @@
+use std::path::PathBuf;
+use test_framework::{Test, TestGroup, TestResult};
+
+fn get_spec(readonly_paths: Vec<String>) -> Spec {
+    SpecBuilder::default()
+        .linux(
+            LinuxBuilder::default()
+                .readonly_paths(readonly_paths)
+                .build()
+                .expect("could not build"),
+        )
+        .build()
+        .unwrap()
+}
+
+fn check_readonly_paths() -> TestResult {
+    // here we abbreviate 'readonly' as ro for variable names,
+    // purely for ease of writing
+
+    let ro_dir = "readonly_dir";
+    let ro_subdir = "readonly_subdir";
+    let ro_file = "readonly_file";
+
+    // in the runtime-tools tests, they start these with a '/',
+    // but in that case, when joined with any path later,
+    // the '/' takes preference, and path is not actually joined
+    // eg : (test).join(t1) = test/t1
+    //      (test).join(.t1) = /t1
+    // which is not what we want, so we leave them without '/'
+    let ro_dir_top = PathBuf::from(ro_dir);
+    let ro_file_top = PathBuf::from(ro_file);
+
+    let ro_dir_sub = ro_dir_top.join(ro_subdir);
+    let ro_file_sub = ro_dir_top.join(ro_file);
+    let ro_file_sub_sub = ro_dir_sub.join(ro_file);
+
+    let root = PathBuf::from("/");
+
+    let ro_paths = vec![
+        root.join(&ro_dir_top).to_string_lossy().to_string(),
+        root.join(&ro_file_top).to_string_lossy().to_string(),
+        root.join(&ro_dir_sub).to_string_lossy().to_string(),
+        root.join(&ro_file_sub).to_string_lossy().to_string(),
+        root.join(&ro_file_sub_sub).to_string_lossy().to_string(),
+    ];
+
+    let spec = get_spec(ro_paths);
+    test_inside_container(spec, &|bundle| {
+        use std::{fs, io};
+        let bundle_path = bundle.as_ref();
+        let test_dir = bundle_path.join(&ro_dir_sub);
+
+        match fs::create_dir_all(&test_dir) {
+            io::Result::Ok(_) => { /*This is expected*/ }
+            io::Result::Err(e) => {
+                bail!(e)
+            }
+        }
+
+        match fs::File::create(test_dir.join("tmp")) {
+            io::Result::Ok(_) => { /*This is expected*/ }
+            io::Result::Err(e) => {
+                bail!(e)
+            }
+        }
+
+        let test_sub_sub_file = bundle_path.join(&ro_file_sub_sub);
+        match fs::File::create(&test_sub_sub_file) {
+            io::Result::Ok(_) => { /*This is expected*/ }
+            io::Result::Err(e) => {
+                bail!(e)
+            }
+        }
+
+        let test_sub_file = bundle_path.join(&ro_file_sub);
+        match fs::File::create(&test_sub_file) {
+            io::Result::Ok(_) => { /*This is expected*/ }
+            io::Result::Err(e) => {
+                bail!(e)
+            }
+        }
+
+        let test_file = bundle_path.join(&ro_file);
+        match fs::File::create(&test_file) {
+            io::Result::Ok(_) => { /*This is expected*/ }
+            io::Result::Err(e) => {
```

Comment:
In the readonly paths test,



The link to the go test is : https://github.com/opencontainers/runtime-tools/blob/master/validation/linux_readonly_paths/linux_readonly_paths.go#L14

What I can't understand is under which conditions/errors in runtime itself is the test supposed to fail : the test sets some paths as read-only in the spec, then sends a fn to validate inside the runtime, and returns error if any from the validation

Now the fn that is sends to inside validate creates the first-dir, sub dirs and files that are supposed to be read-only, and writes some bytes to the files ; and returns err, if any of this returns err. I initially thought that these operations should fail, as the paths are read-only, but as the function is called not from inside, but from outside the container, these operations are valid. Even the (original Go) test itself expect these to be successful. Then I checked the validate-inside fn, but that simply call this fn, then creates and starts the runtime, waits for it to stop and then returns if any error occurred in running of the container. 

So I thought that the test should fail if the read only first are not created, which will fail the creation of container, so I used an empty fn to give to inside-validate ; but that too is passing for runc and youki.

Thus I cannot understand under which circumstances is the test supposed to fail, cause otherwise right now it seems a redundant test.

Basically, as test can only show presence of bugs, not the absence of them, I don't understand the purpose of this test, as I don't understand the condition under which it is supposed to fail.

Reply:
The part you linked does not actually test the writability of the readonly-paths. The runtime test tools start a validation executable within a container in order to test that, see https://github.com/opencontainers/runtime-tools/blob/db715ee5eb891b81a031e680fe92392bff11e31b/cmd/runtimetest/main.go#L939-L962